astropy/cosmology/funcs.py

```diff
@@ -14,7 +21,41 @@
 __doctest_requires__ = {'*': ['scipy']}
 
 
-def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
+def _z_at_array(func, fvals, zmin, zmax, nbins=10000,
+                logspace=True, interpolation=None):
+    """Helper function to interpolate (func, z) over a grid for array input"""
+    if logspace:
+        zgrid = np.logspace(np.log10(zmin), np.log10(zmax), nbins)
+    else:
+        zgrid = np.linspace(zmin, zmax, nbins)
+
+    fgrid = func(zgrid)
+
+    fvals_val = fvals.value
+    fgrid_val = fgrid.to_value(fvals.unit)
+
+    if interpolation is None:
+        if USE_SCIPY:
+            interpolation = 'cubic'
+        else:
+            warnings.warn("""\
+SciPy not found, so falling back to linear numpy interpolation scheme
+for array z_at_value, instead of the normal cubic spline.""")
+            interpolation = 'linear'
+
+    assert interpolation in ['cubic', 'linear']
+
+    if interpolation == 'cubic':
+        interpolator = CubicSpline(fgrid_val, zgrid)
```

Comment:
Just something to think about, but how does ``bc_type`` impact the result? Is there a better default for this use case?

Reply:
Fair enough. Feel free to resolve this thread.

Reply:
More serious point: `CubicSpline` (as probably any standard interpolator requires the independent variable (`fgrid_val`) to be "in strictly increasing order". So for the non-monotonic functions like `angular_diameter_distance` this will require a case distinction, probably setting up the interpolator separately for each side of the maximum and then depend on user input to decide which side is desired.
Maybe for a start, you could enable this _only_ when `zmin`, `zmax` ranges are specified to one side of the maximum only (would become quite intricate anyway if the users wanted solutions to different side for different parts of their input arrays).

Reply:
As it currently stands, non-monotonic functions should cause the interpolator to error, so while this is a serious point that will need to be addressed it currently doesn't get any physics wrong. I think this can be fixed in a future PR.